added compute-filter.cc binary, which computes LP filters per speaker.

pegahgh opened the pull request.

danpovey commented on src/featbin/compute-filter.cc:
```diff
@@ -0,0 +1,122 @@
+    std::string wav_rspecifier = po.GetArg(1),
+      wspecifier = po.GetArg(2);
+    
+    SequentialTableReader<WaveHolder> spk_reader(wav_rspecifier);
```
In general the wave files will be available per-utterance-- to pool the stats over speakers (needed if the spk2utt file is provided), you would need to read the spk2utt file, and iterate over them in a way similar to in compute-cmvn-stats.cc

pegahgh commented:
modified compute-filet.cc by adding --spk2utt option. 
It seems, the Durbin recursion algorithm is not stable and it fails for some speakers.
The problem is that if c < 1e-15, c = 1e-15  and c is getting negative in some cases. E is scaled using new c and E is getting very small and we scale k by 1/E!

danpovey commented on src/featbin/compute-filter.cc:
```diff
@@ -0,0 +1,223 @@
+            }
+          }
+          SubVector<BaseFloat> waveform(wav_data.Data(), this_chan);
+          int32 corr_win_size = waveform.Dim() - lpc_order;
```
don't call this corr_win_size since you use corr_window_size elsewhere- better be consistent

danpovey replied:
Actually, don't pass in corr_win_size at all.  You just have it equal to waveform.Dim() - lpc_order-- the resizing inside the function with kCopyData would not be necessary if, inside the function, you could assume it equalled this value all the time.

pegahgh replied:
I thought by passing corr_win_size, we can have handle unequal length of inputs  for correlation functions corr(x,y) , where len(x) > len(y), and we can handle it by padding zero to y, which is equal to padding zero to original waveform.
I will remove option and assume len(x) = len(y) in corr(x,y)

danpovey commented on src/featbin/compute-filter.cc:
```diff
@@ -0,0 +1,223 @@
+          spk_mean(num_chan, 0);
+        // Accumulate waveform sum and count over all utterance in speaker
+        // to compute mean of waveform for whole speaker over all channels.
+        if (use_mean) {
```
no reason to compute the mean for each channel, since computation is done only for one channel.
Work out the mean inside the 'AccStatsForCorrelation' function instead.

danpovey commented on src/featbin/compute-filter.cc:
```diff
@@ -0,0 +1,223 @@
+  (*norms)(0) += e1;
+  (*inner_prod)(0) += e1;
+  for (int32 lag = 1; lag < lpc_order; lag++) {
+    SubVector<BaseFloat> sub_vec2(zero_mean_wave, lag, corr_window_size);
```
this way of computing e2 is very inefficient, as it only differs slightly from the previous e2.  better to just add the single-sample corrective terms.

danpovey commented:
Regarding the Durbin recursion algorithm, I think in principle it's
supposed to be always stable, that is part of the reason it's used.  I
suspect that you have made some mistake in accumulating the stats.
Try applying the learned filter (in the 'whitening' mode) and verify that
if you apply your compute-filter code again, it learns a filter that's just
1 0 0 0.  (or 0 0 0 0 if you assume the first element is 1).


On Mon, Feb 29, 2016 at 2:14 PM, pegahgh <notifications@github.com> wrote:

> modified compute-filet.cc by adding --spk2utt option.
> It seems, the Durbin recursion algorithm is not stable and it fails for
> some speakers.
> The problem is that if c < 1e-15, c = 1e-15 and c is getting negative in
> some cases. E is scaled using new c and E is getting very small and we
> scale k by 1/E!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/danpovey/kaldi/pull/20#issuecomment-190339761>.
>

pegahgh commented:
Yes, Durbin algorithm is designed to be stable, since |ki| < 1. The point
is that in Durbin algorithm implementation in kaldi, the if condition is
added where
if (c < 1e-5) c = 1e-5.
When I removed this if condition, the algorithm gets stable!!!

On Mon, Feb 29, 2016 at 3:14 PM, Daniel Povey <notifications@github.com>
wrote:

> Regarding the Durbin recursion algorithm, I think in principle it's
> supposed to be always stable, that is part of the reason it's used. I
> suspect that you have made some mistake in accumulating the stats.
> Try applying the learned filter (in the 'whitening' mode) and verify that
> if you apply your compute-filter code again, it learns a filter that's just
> 1 0 0 0. (or 0 0 0 0 if you assume the first element is 1).
>
>
> On Mon, Feb 29, 2016 at 2:14 PM, pegahgh <notifications@github.com> wrote:
>
> > modified compute-filet.cc by adding --spk2utt option.
> > It seems, the Durbin recursion algorithm is not stable and it fails for
> > some speakers.
> > The problem is that if c < 1e-15, c = 1e-15 and c is getting negative in
> > some cases. E is scaled using new c and E is getting very small and we
> > scale k by 1/E!
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/danpovey/kaldi/pull/20#issuecomment-190339761>.
>
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/danpovey/kaldi/pull/20#issuecomment-190364965>.
>

danpovey commented:
I checked the HTK source and this line does not come from HTK.  It looks
like someone added that statement during the early days of Kaldi, before
the official release, and there seems to be no info on who it was (it
doesn't look like my code, e.g. the misspelling and lack of explanation).
If this code is changed, it's possible that PLP coefficients will change
and become incompatible.
What kind of data is failing this, and which coefficient are you getting
the errors on ?

Dan


On Mon, Feb 29, 2016 at 4:13 PM, pegahgh <notifications@github.com> wrote:

> Yes, Durbin algorithm is designed to be stable, since |ki| < 1. The point
> is that in Durbin algorithm implementation in kaldi, the if condition is
> added where
> if (c < 1e-5) c = 1e-5.
> When I removed this if condition, the algorithm gets stable!!!
>
> On Mon, Feb 29, 2016 at 3:14 PM, Daniel Povey <notifications@github.com>
> wrote:
>
> > Regarding the Durbin recursion algorithm, I think in principle it's
> > supposed to be always stable, that is part of the reason it's used. I
> > suspect that you have made some mistake in accumulating the stats.
> > Try applying the learned filter (in the 'whitening' mode) and verify that
> > if you apply your compute-filter code again, it learns a filter that's
> just
> > 1 0 0 0. (or 0 0 0 0 if you assume the first element is 1).
> >
> >
> > On Mon, Feb 29, 2016 at 2:14 PM, pegahgh <notifications@github.com>
> wrote:
> >
> > > modified compute-filet.cc by adding --spk2utt option.
> > > It seems, the Durbin recursion algorithm is not stable and it fails for
> > > some speakers.
> > > The problem is that if c < 1e-15, c = 1e-15 and c is getting negative
> in
> > > some cases. E is scaled using new c and E is getting very small and we
> > > scale k by 1/E!
> > >
> > > —
> > > Reply to this email directly or view it on GitHub
> > > <https://github.com/danpovey/kaldi/pull/20#issuecomment-190339761>.
> >
> > >
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/danpovey/kaldi/pull/20#issuecomment-190364965>.
>
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/danpovey/kaldi/pull/20#issuecomment-190398910>.
>

danpovey commented on src/featbin/compute-filter.cc:
```diff
@@ -0,0 +1,223 @@
+    e2 = VecVec(sub_vec2, sub_vec2);
+    sum = VecVec(sub_vec1, sub_vec2);
+    (*inner_prod)(lag) += sum;
+    (*norms)(lag) += e2;
```
Actually I don't think you need to compute these norms at all.
From here https://en.wikipedia.org/wiki/Autocorrelation
it says "In signal processing, the above definition is often used without the normalization, that is, without subtracting the mean and dividing by the variance. When the autocorrelation function is normalized by mean and variance, it is sometimes referred to as the autocorrelation coefficient".
It looks to me from feature-plp.cc that the autocorrelation coefficients (not autocorrelations) are used as the input to the filter computation.  Generally they will be much larger than 1 because the signal is not divided by 2^16.  This means that it won't tend to hit that floor.
I am not at all happy with that floor of 1e-5 in the code but removing it will require testing.

danpovey commented on src/featbin/compute-filter.cc:
```diff
@@ -0,0 +1,223 @@
+
+  Vector<BaseFloat> zero_mean_wave(wave);
+  if (mean != 0.0) 
+    zero_mean_wave.Add(mean);
```
do you mean Add(-mean)?

danpovey replied:
plus I don't see a reason to pass the mean in; just compute it inside the function.
